Improve assign ui

emcelroy opened the pull request:
This is related to the changes in this portal/rigse PR: https://github.com/concord-consortium/rigse/pull/721

emcelroy commented:
@scytacki OK. I did find a couple problems though that I'll try to fix today. 

If you click the assign button while anonymous, then click the dialog's login button and log in, you get taken back to the STEM Resource finder with the lightbox and assign dialog open. That much is good, but if you close those things and click on another resource, the assign dialog opens automatically. And when you close that resource's lightbox, the URL changes to the previous resource's lightbox URL instead of the site's root URL.

scytacki commented:
The first issue could probably be fixed by removing the openAssign from the savedUrl.
The second issue is going to be hard. I checked and it is happen on production too.
If you open a lightbox, reload the page. Then close the lightbox, open another resource, and close it. Now your window location will be pointing to the first lightbox.

I suspect the second issue happens on production right now in some form. If you visit a

scytacki commented:
I'm not sure, but I think it would be better to store the openAssign, savedUrl and savedTitle in the state. And use the getInitialState method to add these to the state. 
I think getDefaultProps will be called just once for for the class, and the class will be reused by each lightbox. But you might want to check that with some logging or debug statements before making changes.

emcelroy commented:
@scytacki What do you think of the latest changes? It all works as expected now. There's at least one part that's still kind of kludgy -- lines 65-71. I couldn't think of a better way to handle that URL problem without requiring a much more significant change.

scytacki commented on src/library/components/resource-lightbox.js:
```diff
@@ -59,6 +57,19 @@ var ResourceLightbox = Component({
+      // possibility the lightbox is being shown on a collection page.
+      savedUrl = '/'
+    }
+
```
I'm not sure this is going to work very well. I looked into why the url is the root in this case:
- user loads url that opens a lightbox directly like: https://learn.concord.org/resources/799/3d-exploration-of-bound-antibody-and-antigen
- then the user closes the lightbox
- now the browser url is the root

This happens because of this line: https://github.com/concord-consortium/rigse/blob/a7eb234951ef2ea0d9add4d6f575ce001be41e9b/app/views/home/home.html.haml#L29

With the switching to state that property setting in the portal code will have no effect. 

The redirect check above would also not pick up a direct load of a resource. So with this new code above I'd expect:
- user loads url that opens a lightbox directly like: https://learn.concord.org/resources/799/3d-exploration-of-bound-antibody-and-antigen
- then the user closes the lightbox
- now the browser url is still the lightbox url (no change)

 A quick thought is to add a `savedUrl` parameter to the URL. If the `savedUrl` parameter isn't set then use `/`.  So now on a collection page the savedUrl parameter would be set to the collection page.  If the user chooses to login then they will be redirected back to the resource lightbox with this same savedUrl parameter so when they close the page they will still be on the collection page. The downside of this is that the URL for a resource lightbox on a collection page will be different than the same lightbox on the stem resource finder. So if someone shares it they  have this extra param in it. That might be OK or a feature, or it could be problematic if you want all sharing to be to a single place.  If you do implement this I'd suggest changing the name from savedUrl to background or something that is more clear that this is the url underneath.  It should also forced to be a just a path, not a full URL. It is insecure to let people construct URLs that redirect to other sites.

emcelroy replied:
@scytacki I've made some changes, but I still feel like things are still a little kludgy and getting more complicated in that regard, too. There are some accompanying portal code changes at https://github.com/concord-consortium/rigse/pull/724

For what it's worth, I did test out the previous work and it all seemed to function properly, including in the second situation you described. From what I could tell, it didn't seem to introduce any new bugs concerning the URLs.

The latest changes fix at least some of those URL bugs though, again, I feel like the fixes are iffy. I suspect it's the case that I should stop trying to make this work with things as they are  and start moving everything to portal-pages.

scytacki replied:
It seems like we should work on this together.  I don't want to make it more complicated. If it was working with your code from before then I'm confused so I'd like to understand it.

scytacki commented on src/library/library.js:
```diff
@@ -33,6 +33,7 @@ const renderComponentFn = function (ComponentClass) {
 
 window.PortalPages = {
   settings: {}, // default to empty, used to set flags from portal templates
+  initialPath: window.location.pathname,
```
A comment here would be useful:
> The URL of the page will change as the user opens resource lightboxes. This initialPath provides a way for the code to know the initial page loaded was the stem resource finder, a collection page, or a direct resource url.

scytacki commented on src/library/components/resource-lightbox.js:
```diff
@@ -31,18 +38,29 @@ var ResourceLightbox = Component({
+    document.title = this.state.savedTitle
     try {
-      window.history.replaceState({}, document.title, this.props.savedUrl)
+      if (this.state.parentPage !== '/' && this.state.parentPage !== PortalPages.initialPath) {
```
Another comment here:
> When the parentPage is not `/` and it doesn't match the initiaPath then reload the page
> This can happen when a resource URL is opened directly. The parentPage URL parameter can be set to be a collection page but the initialPath will be the direct resource URL. 
> If the lightbox was opened from a collection page, the initialPath will be the collection page and the parentPage will be the collection page, so then we can just update the URL and close the lightbox.

emcelroy commented:
@scytacki Thanks. I've added all those comments. 

Looks like `getParentPageType` is no longer used. I had added it here https://github.com/concord-consortium/portal-pages/pull/117 but then modified things later such that it wasn't needed anymore, but forgot to remove it. I've removed it in the latest changes.

emcelroy commented on src/library/components/resource-lightbox.js:
```diff
@@ -31,18 +38,29 @@ var ResourceLightbox = Component({
     } catch (e) {}
+
     jQuery('html, body').css('overflow', 'auto')
     jQuery('.home-page-content').removeClass('blurred')
```
@scytacki I made this change. I also added a line that makes the page look blank immediately after the lightbox closes so that the reload is even less confusing/noticeable.

emcelroy commented:
I meant to add that I'm fine with the URL format as is for now. I agree your suggested change would improve things, but I think it'd be best to get these changes out soon and work on that later. I'd also like to find out what the implications for having the same content at two different URLs would be for SEO. I'm pretty sure there's a way to prevent it from being a problem, but I'll have to refresh my memory on that.
